Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+}
+
+Vec4 calcJacobianOfNextPathError(
+    const CurveSegment& curve,
+    const State& s,
+    const LineSegment& line,
+    const UnitVec3& axis)
+{
+    const Transform& X_GQ = curve.getDataPos(s).X_GQ;
+
+    const CurveSegmentData::Instance& dataInst = curve.getDataInst(s);
+    const Real a                               = dataInst.jacobi_Q[0];
+    const Real r                               = dataInst.jacobi_Q[1];
+
+    // Partial derivative of path error to contact point position (x_QS),
+    // represented in the frenet frame.
+    Vec3 dErrDx = axis - line.direction * dot(line.direction, axis);
+    dErrDx      = X_GQ.RInv() * (-dErrDx / line.length);
+
+    return {dErrDx[0], dErrDx[2] * a, dErrDx[2] * r, dErrDx[0]};
+}
+
+Vec4 calcJacobianOfPathErrorAtP(
+    const CurveSegment& curve,
+    const State& s,
+    const LineSegment& line,
+    const UnitVec3& axis)
+{
+    const Transform& X_GP = curve.getDataPos(s).X_GP;
+
+    const CurveSegmentData::Instance& dataInst = curve.getDataInst(s);
+
+    const Real tau = dataInst.torsion_P;
+    const Real kt  = dataInst.curvatures_P[0];
+    const Real kb  = dataInst.curvatures_P[1];
+
+    // Compute partial derivative of path error to the natural geodesic
+    // corrections. For derivation see TODO.
+    const Vec3 y = X_GP.RInv() * cross(axis, line.direction);
+    return Vec4{tau * y[0] + kt * y[2], -kb * y[0] - tau * y[2], -y[1], 0.} +
+           calcJacobianOfPrevPathError(curve, s, line, axis);
+}
+
+Vec4 calcJacobianOfPathErrorAtQ(
+    const CurveSegment& curve,
+    const State& s,
+    const LineSegment& line,
+    const UnitVec3& axis)
+{
+    const Transform& X_GQ = curve.getDataPos(s).X_GQ;
+
+    const CurveSegmentData::Instance& dataInst = curve.getDataInst(s);
+
+    const Real tau = dataInst.torsion_Q;
+    const Real kt  = dataInst.curvatures_Q[0];
+    const Real kb  = dataInst.curvatures_Q[1];
+
+    const Real a = dataInst.jacobi_Q[0];
+    const Real r = dataInst.jacobi_Q[1];
+
+    const Real aDot = dataInst.jacobiDot_Q[0];
+    const Real rDot = dataInst.jacobiDot_Q[1];
+
+    // Partial derivative of path error to the given axis, represented in the
+    // frenet frame.
+    const Vec3 y = X_GQ.RInv() * cross(axis, line.direction);
+
+    return Vec4{
+               tau * y[0] + kt * y[2],
+               -a * kb * y[0] - aDot * y[1] - a * tau * y[2],
+               -r * kb * y[0] - rDot * y[1] - r * tau * y[2],
+               tau * y[0] + kt * y[2]} +
+           calcJacobianOfNextPathError(curve, s, line, axis);
+}
+
+template <size_t N>
+void calcPathErrorJacobian(
+    const CableSpan::Impl& cable,
+    const State& s,
+    const std::vector<LineSegment>& lines,
+    std::array<CoordinateAxis, N> axes,
+    Matrix& J)
+{
+    // Number of free coordinates for a generic geodesic.
+    constexpr int NQ = c_GeodesicDOF;
+
+    const int numberOfCurvesInContact = lines.size() - 1;
+
+    // Reset the values in the jacobian.
+    J *= 0.;
```

Comment:
Is this truly resetting the values? Multiplication by zero doesn't reset NaNs or sign-bits, I'd be surprised if there was no way of actually assigning a new value to all elements of a `Matrix`.

Reply:
ah yes: `Matrix::setToZero()`